### DISCUSSION TITLE
Whisper is not scale invariant

### DISCUSSION THREAD

**Discussion by sanchit-gandhi** [openai/whisper#428] [Q&A] [answered]
The [Whisper paper](https://cdn.openai.com/papers/whisper.pdf) states that the inputs are scaled to a range [-1, 1] with approximately zero mean:
<img width="450" alt="Screenshot 2022-10-27 at 17 01 54" src="https://user-images.githubusercontent.com/93869735/198351777-ee7a9a34-dbef-4947-a3fd-391dad72d14b.png">

Is this feature normalisation performed on the raw audio inputs, on on the log-Mel spectrograms? The reason that I ask is because the Whisper model does **not** give the same outputs for the same audio inputs that are scaled/re-scaled.

If we take a sample from the LibriSpeech dataset, scale it to a range of [0, 65535], and pass it through the Whisper model, we get a junk output. Note that the value 65535 is chosen arbitrarily here, just to demonstrate what we get for an audio sample with a large dynamic range.

If take this same input, re-scale it to a range [-1, 1] and zero mean, and pass it through the Whisper model, we get a meaningful transcription. Consequently, the Whisper model is **not** scale invariant. This makes me question whether feature normalisation is being performed correctly in the Whisper model?

Install necessary packages:
```
pip install git+https://github.com/openai/whisper.git
pip install datasets
```

```python
import torch
import whisper
from datasets import load_dataset

device = "cuda" if torch.cuda.is_available() else "cpu"

model = whisper.load_model("base.en")
model.to(device)

# define the decoding options
options = whisper.DecodingOptions(language="en", without_timestamps=True, fp16=False)

# load an audio sample from LS val-clean (dataset is ~9MB total)
ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
audio = ds[0]["audio"]["array"]

# prepare the (unscaled) audio input
audio = torch.tensor(audio)
audio = ((audio - audio.min()) / (audio.max() - audio.min())) * 65535  # Rescale to [0, 65535] to demonstrate issue

def inference(audio):
    audio = whisper.pad_or_trim(audio.flatten())
    mel = whisper.log_mel_spectrogram(audio)
    results = model.decode(mel, options)
    return results.text

unscaled_transcription = inference(audio)

# Rescale audio to [-1, 1] and then to zero mean
audio_scaled = ((audio - audio.min()) / (audio.max() - audio.min())) * (2) - 1  
audio_scaled = audio_scaled - audio_scaled.mean()

scaled_transcription = inference(audio_scaled)

print("Target:   ", ds[0]["text"])
print("Unscaled: ", unscaled_transcription)
print("Scaled:   ", scaled_transcription)
```

```
Target:    MISTER QUILTER IS THE APOSTLE OF THE MIDDLE CLASSES AND WE ARE GLAD TO WELCOME HIS GOSPEL
Unscaled:  I
Scaled:    Mr. Quilter is the apostle of the middle classes, and we are glad to welcome his gospel.
```

We can see that the transcription for the audio samples in the range [0, 65535] is entirely wrong. The transcription for the audio samples in the range [-1, 1] is transcribed perfectly. Is the Whisper model correctly normalising the audio inputs, whether that be in the audio space or log-Mel space?

Original issue: https://github.com/huggingface/transformers/issues/19888 (credits [JeffreyWardman](https://github.com/JeffreyWardman))
cc @jongwook

**Comment by jongwook** [comment, marked as answer]
The model wasn't intended to be scale-invariant. "For feature normalization" in the quoted paragraph meant that the log magnitude values in the Mel spectrogram have been manually rescaled to (roughly) [-1, 1] during training. This means that the training data was always in this range, and the model will struggle for inputs outside the range.

What's applied during training is equivalent to the below, which didn't use exact global statistics of the data but used the fact that the log Magnitudes mostly ranges from -80 dB to 0 dB. We could update the paper to clarify this.

https://github.com/openai/whisper/blob/f68057001688ae75e59c168019850bc0f2a5e742/whisper/audio.py#L114-L124

For inference, we didn't add any normalization layer (although I see how it could make sense). The model will be somewhat invariant to normal volume changes, but apparently not for a 32768x change.

**Reply by sanchit-gandhi** [reply]
I see! Thank you for the clarification @jongwook, it's most helpful indeed 🤗 The 32768x change is extreme and purely for demonstration purposes 😉 It takes a very large dynamic range to push the normalised log-Mel spectrogram out of the normal operating values. 

Perhaps it would make sense to provide an (optional) argument to the feature-extractor indicating whether the audio inputs should be normalised in the audio space:
* `do_normalise` (Optional, defaults to `False`): whether or not to normalise the audio inputs prior to computing the log-Mel features

In Transformers, this would look something along the lines of:
```python
from transformers import WhisperFeatureExtractor

feature_extractor = WhisperFeatureExtractor.from_pretrained("openai/whisper-base.en")
# don't normalise
input_features = feature_extractor(audio, do_normalise=False).input_features[0]
# do normalise
input_features = feature_extractor(audio, do_normalise=True).input_features[0]
```
-> we can add this quite easily for more control over inference
